src/Runtime/Loop.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Runtime;
+
+use Twig\Error\RuntimeError;
+
+/**
+ * Represents a for loop variable.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class Loop implements \Iterator
+{
+    private \Iterator $seq;
+    private bool $iterated;
+    private int $index0;
+    private int $revindex0;
+    private bool $first;
+    private bool $last;
+    private int $length;
+
+    public function __construct($seq, private $parent)
+    {
+        $this->seq = is_iterable($seq) ? (is_array($seq) ? new \ArrayIterator($seq) : $seq) : new \ArrayIterator([]);
+        $this->rewind();
+    }
+
+    public function current(): mixed
+    {
+        return $this->seq->current();
+    }
+
+    public function key(): mixed
+    {
+        return $this->seq->key();
+    }
+
+    public function next(): void
+    {
+        $this->seq->next();
+
+        $this->iterated = true;
+        ++$this->index0;
+        $this->first = false;
+        if (isset($this->length)) {
+            --$this->revindex0;
+            $this->last = 0 === $this->revindex0;
+        }
+    }
+
+    public function rewind(): void
+    {
+        $this->seq->rewind();
+
+        $this->iterated = false;
+        $this->index0 = 0;
+        $this->first = true;
+        if (is_array($this->seq) || (is_object($this->seq) && $this->seq instanceof \Countable)) {
```

Comment:
`is_object` is useless. The `instanceof` operator will return `false` for non-objects

Reply:
Right, now that I always convert to an \Iterator, I can simplify this a lot.